source_modelling/fsp.py

```diff
@@ -0,0 +1,250 @@
+    nz : float or None
+        The number of subfaults along dip.
+    dx : float
+        The length of each subfault.
+    dz : float
+        The width of each subfault.
+    fmin : float or None
+        The minimum frequency in the inversion.
+    fmax : float or None
+        The maximum frequency in the inversion.
+    time_window_count : float or None
+        The slip (or rake) time windows for each subfault.
+    time_window_length : float or None
+        The time window length for each subfault.
+    time_shift : float or None
+        The time shift for each time window.
+    segment_count : int
+        Number of segments (subfaults).
+    data : pd.DataFrame
+        The data frame of info on each subfault.
+    """
+
+    event_tag: str
+
+    # Hypocentre parameters
+    latitude: float
+    longitude: float
+    depth: float
+    hypx: Optional[float]
+    hypz: Optional[float]
+
+    # Fault and parameters
+    length: float
+    width: float
+    strike: float
+    dip: float
+    rake: float
+    htop: float
+
+    # Rupture parameters
+    magnitude: float
+    moment: float
+    average_rise_time: Optional[float]
+    average_rupture_speed: Optional[float]
+    slip_velocity_function: str
+
+    # Model subfault sizes
+    nx: Optional[int]
+    nz: Optional[int]
+    dx: float
+    dz: float
+
+    # Inversion parameters
+    fmin: Optional[float]
+    fmax: Optional[float]
+
+    # Time window parameters
+    time_window_count: Optional[float]
+    time_window_length: Optional[float]
+    time_shift: Optional[float]
+
+    # Number of segments (== len(data))
+    segment_count: int
+    data: pd.DataFrame
+
+    @classmethod
+    def read_from_file(cls: Callable, fsp_ffp: Path) -> "FSPFile":
+        """Parse an FSPFile.
+
+        Parameters
+        ----------
+        fsp_ffp : Path
+            Path to the FSP file.
+
+        Returns
+        -------
+        FSPFile
+            The parsed FSPFile.
+        """
+        with open(fsp_ffp, "r") as fsp_file_handle:
+            header = ""
+            # Collect and normalise all the lines that make up the header
+            # of the file.
+            for line in fsp_file_handle:
+                if line.startswith("% Data :"):
+                    break
+                if (
+                    line.startswith("% -")
+                    or line.strip() == "%"
+                    or line.startswith("% Event :")
+                ):
+                    continue
+                # Strip the leading "% ", and deduplicate the spaces in the line.
+                # This is required to normalise the string so that the parse
+                # module can handle the rest of the parsing
+                header += " ".join(line.lstrip("% ").split()) + "\n"
+
+            # Now we can parse the header according to the header pattern.
+            parse_result = parse.parse(
+                HEADER_PATTERN,
+                header,
+            )
```

Comment:
I like this

Reply:
Yeah the parse package is really nice, good alternative to regex for simple use cases.

Reply:
Finally we have a `scanf ` alternative.